src/test/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/ClusterMonitorSpec.scala

```diff
@@ -0,0 +1,245 @@
+    googleId = UUID.randomUUID(),
+    googleProject = "dsp-leo-test",
+    googleServiceAccount = "not-a-service-acct@google.com",
+    googleBucket = "bucket1",
+    clusterUrl = Cluster.getClusterUrl("dsp-leo-test", "name1"),
+    operationName = "op1",
+    status = ClusterStatus.Creating,
+    hostIp = None,
+    createdDate = Instant.now(),
+    destroyedDate = None,
+    labels = Map("bam" -> "yes", "vcf" -> "no"))
+
+  val deletingCluster = Cluster(
+    clusterName = "name2",
+    googleId = UUID.randomUUID(),
+    googleProject = "dsp-leo-test",
+    googleServiceAccount = "not-a-service-acct@google.com",
+    googleBucket = "bucket1",
+    clusterUrl = Cluster.getClusterUrl("dsp-leo-test", "name1"),
+    operationName = "op1",
+    status = ClusterStatus.Deleting,
+    hostIp = Some("numbers.and.dots"),
+    createdDate = Instant.now(),
+    destroyedDate = None,
+    labels = Map("bam" -> "yes", "vcf" -> "no"))
+
+  /**
+    * Extends ClusterMonitorSupervisor so the akka TestKit can watch the child ClusterMontitorActor's.
+    */
+  class TestSupervisor(gdDAO: DataprocDAO, dbRef: DbReference) extends ClusterMonitorSupervisor(MonitorConfig(1 second), gdDAO, dbRef) {
+    override def createChildActor(cluster: Cluster): ActorRef = {
+      val child = super.createChildActor(cluster)
+      testKit watch child
+      child
+    }
+  }
+
+  def createClusterSupervisor(dao: DataprocDAO): ActorRef = {
+    system.actorOf(Props(new TestSupervisor(dao, DbSingleton.ref)))
+  }
+
+  // Pre:
+  // - cluster exists in the DB with status Creating
+  // - dataproc DAO returns status RUNNING
+  // Post:
+  // - cluster is updated in the DB with status Running and the host IP
+  // - monitor actor shuts down
+  "ClusterMonitorActor" should "monitor until RUNNING state" in isolatedDbTest {
+    dbFutureValue { _.clusterQuery.save(creatingCluster) } shouldEqual creatingCluster
+
+    val dao = mock[DataprocDAO]
+    when {
+      dao.getClusterStatus(eqq(creatingCluster.googleProject), eqq(creatingCluster.clusterName))(any[ExecutionContext])
+    } thenReturn Future.successful(ClusterStatus.Running)
+
+    when {
+      dao.getClusterMasterInstanceIp(eqq(creatingCluster.googleProject), eqq(creatingCluster.clusterName))(any[ExecutionContext])
+    } thenReturn Future.successful(Some("1.2.3.4"))
+
+    createClusterSupervisor(dao) ! ClusterCreated(creatingCluster)
+
+    expectMsgClass(5 seconds, classOf[Terminated])
+    val updatedCluster = dbFutureValue { _.clusterQuery.getByName(creatingCluster.googleProject, creatingCluster.clusterName) }
+    updatedCluster shouldBe 'defined
+    updatedCluster.map(_.status) shouldBe Some(ClusterStatus.Running)
+    updatedCluster.flatMap(_.hostIp) shouldBe Some("1.2.3.4")
+  }
+
+  // Pre:
+  // - cluster exists in the DB with status Creating
+  // - dataproc DAO returns status CREATING, UNKNOWN, or UPDATING
+  // Post:
+  // - cluster is not changed in the DB
+  // - monitor actor does not shut down
```

Comment:
another test: dataproc DAO returns one of the "halfway" responses, like running but no IP, or error but no details

Reply:
👍

Reply:
Done, see below 2 tests